pkg/controller/registry/reconciler/grpc.go

```diff
@@ -180,62 +193,54 @@ func (c *GrpcRegistryReconciler) ensurePod(source grpcCatalogSourceDecorator, ov
 	return nil
 }
 
-// ensureUpdatePod checks that for the same catalog source version the same imageID is running
+// ensureUpdatePod checks that for the same catalog source version the same container imageID is running
 func (c *GrpcRegistryReconciler) ensureUpdatePod(source grpcCatalogSourceDecorator) error {
 	if !source.Poll() {
 		return nil
 	}
-	var updateFlag bool
+
 	currentLivePods := c.currentPods(source)
 	currentUpdatePods := c.currentUpdatePods(source)
 
-	for _, updatePod := range currentUpdatePods {
-		if updatePodByDigest(updatePod, currentLivePods) {
-			logrus.WithField("CatalogSource", source.GetName()).Info("detect image update for catalogsource pod")
-
-			updateFlag = true
-			updatePod.Labels[CatalogSourceLabelKey] = source.GetName()
-			updatePod.Labels[CatalogSourceUpdateKey] = ""
-
-			// Update the update pod to promote it to serving pod
-			_, err := c.OpClient.KubernetesInterface().CoreV1().Pods(source.GetNamespace()).Update(context.TODO(), updatePod, metav1.UpdateOptions{})
-			if err != nil {
-				return errors.Wrapf(err, "error creating new pod: %s", source.Pod().GetName())
-			}
-
-			break
+	if source.Update() && len(currentUpdatePods) == 0 {
```

Comment:
Is there ever a time where the source is ready for an update but the number of currentUpdatePods is greater than 1? Could this happen if there was a CatSrc with a bad image?

Reply:
This is a scenario we definitely want to avoid and I think we have it in our code currently if the polling interval is really small. Basically with this additional check we only create an update pod if there is not one present on the cluster. 

If the catsrc has a bad image, it never reports ready (so we can't detect the imageID of the update pod). That sounds like an additional case that we should handle  (but do not currently). I can open up another BZ to handle this case.

Reply:
Thanks for opening the BZ, marking this resolved.